src/RECONR/ReferenceGrid/src/call.hpp

```diff
@@ -1,65 +1,172 @@
+  energies.push_back( upperLimit );
+
+  if ( not ranges::is_sorted( energies ) ){
+    orlp::pdqsort( energies.begin(), energies.end() );
+  }
+
+  energies |= ranges::action::unique;
+  return energies;
 }
 
-template< typename resolvedL >
-auto operator()( const resolvedL& lValue,
-                 std::conditional_t< true, void*,  decltype( lValue.resonances() ) > = nullptr) const {
+auto operator()( const unresolved::CaseA& caseA ) const {
+  std::vector< double > energies;
+  const auto lowerLimit = caseA.EL();
+  const auto upperLimit = nudgeDown( caseA.EH() );
+  energies.reserve( std::ceil( 5.645828264742274 *
+                               std::log( upperLimit / lowerLimit ) ) );
+  energies.push_back( lowerLimit );
+  fill( lowerLimit, upperLimit, energies );
+  return energies;
+}
+
+auto operator()( const unresolved::CaseB& caseB ) const {
+  std::vector< double > energies;
+  const auto lowerLimit = caseB.EL();
+  const auto upperLimit = nudgeDown( caseB.EH() );
+  energies.reserve( 8 * std::ceil( std::log( upperLimit / lowerLimit ) ) );
+  energies.push_back( lowerLimit );
+  auto clampedES =
+    caseB.ES()
+    | ranges::view::drop_while
+      ( [=]( auto energy ){ return energy <= lowerLimit; } )
+    | ranges::view::take_while
+      ( [=]( auto energy ){ return energy < upperLimit; } );
 
-  return lValue.resonances()
-    | ranges::view::for_each( *this );
+  RANGES_FOR( const auto energy, clampedES ){
+    fill( energies.back(), energy, energies );
+  }
+
+  fill( energies.back(), upperLimit, energies );
+  return energies;
 }
 
-// template< typename URR, 
-//           typename =decltype( std::declval< URR >().jValues() ) >
-// auto operator()( const URR& lValues ) const {
-// }
-
-template< typename T >
-auto operator()( const T& range,
-                 std::conditional_t< true, void*,  decltype( std::declval< T >().lValues() ) > = nullptr) const {
-  auto energies = range.lValues()
-    | ranges::view::for_each( *this )
-    // | ranges::view::drop_while(
-    //     [&]( const auto& E ){ return E < range.EL(); } );
-    // | ranges::view::take_while(
-    //     [&]( const auto& E ){ return E < range.EH(); } )
-    | ranges::to_vector;
-  // echo< decltype( e0 ) > {};
+template< typename UnresolvedLvalue,
+          if_< hasJvalues< UnresolvedLvalue > > = true >
+auto
+operator()( const UnresolvedLvalue& lValue ) const {
+  return
+    lValue.jValues()
+    | ranges::view::transform
+      ( []( const auto& jValue ){ return jValue.ES(); } );
+}
+
+auto operator()( const unresolved::CaseC& caseC ) const {
+  std::vector< double > energies;
 
-  // energies.push_back( range.EL() );
-  // energies.push_back( range.EH() );
+  const auto lowerLimit = caseC.EL();
+  const auto upperLimit = nudgeDown( caseC.EH() );
+
+  std::vector< double > firstPass;
+  firstPass.push_back( lowerLimit );
+
+  {
+    auto jValues = caseC.lValues() | ranges::view::for_each( *this );
```

Comment:
Concatenating the filtered jvalue ranges together then sorting is naive from a complexity point of view, given each jvalue range is already sorted. Merging and sorting using a prioriity queue would be preferable.

That said, I wouldn't expect much of a difference, practically speaking, after accounting for pdqsort's special considerations for mostly sorted ranges.

Reply:
Note that NJOY2016 uses the concatenate and sort approach as well, although there are some important differences. 

1. NJOY2016 eagerly fills each jvalues, which, given the predetermined fill points, amounts to a lot of redunant points.
1. While this sorts at each resonance range (enabling a cheap sort-free merge at the end), NJOY2016 defers the sorting step until after the file3, resolved, and unresolved ranges are concatenated.
1. Here we're using pdqsort, which has `O(N)` complexity for mostly sorted lists. NJOY2016 is using selection sort, which has `O(N^2)` complexity in the best case. 
1. We're using a seperately allocated block of memory for each range whereas NJOY2016 uses a single (fixed size) allocation. This costs a bit more up front but more than pays for the extra allocation time come merge time.

Points [2] and [3] (and to a lesser extent [1] ) taken together are a bit troubling, given, in NJOY2016, sorting has been defered as long as possible. `O(N^2)` gets expensive fast. Might be worthwhile to take a RECONR call on U-238 into gprof and see how much time is spent in the `order` subroutine.